### DISCUSSION TITLE
Logging response content when streaming

### DISCUSSION THREAD

**Discussion by mkiami-accenture** [mitmproxy/mitmproxy#6996] [Q&A]
Hello,

From the documentation, the response event mentions "**If response streaming is active, this event fires after the entire body has been streamed**", and the following is mentioned when data is streamed "Stream data to the client if response body exceeds the given threshold. **If streamed, the body will not be stored in any way.** Understands k/m/g suffixes, i.e. 3m for 3 megabytes."

We have a requirement to log response data, including streaming responses.

So far, logging content for non-streaming responses is possible by capturing it with the response event. When streaming is enabled with flow.response.stream = True, then the content is empty.

- Is there any option that enables response content logging out-of-the-box with mitmproxy?
- If not, what part do you suggest we modify to make this possible?

Thank you for your support.

**Comment by mhils** [comment]
> * Is there any option that enables response content logging out-of-the-box with mitmproxy?

Currently no.

> * If not, what part do you suggest we modify to make this possible?

The best approach is probably to add a new `store_streamed_bodies` option, and then additionally put the body into the receive buffer in the request/response stream handlers: https://github.com/mitmproxy/mitmproxy/blob/5496045e0bf9aa7276558416bda7c8db2c217345/mitmproxy/proxy/layers/http/__init__.py#L304-L344.

On EndOfMessage you can then do this: 
https://github.com/mitmproxy/mitmproxy/blob/5496045e0bf9aa7276558416bda7c8db2c217345/mitmproxy/proxy/layers/http/__init__.py#L358-L359


Contributions are welcome! :)

**Reply by mkiami** [reply]
Hello,

Thank you for your response. 

After looking into other options, I think changing mitmproxy to buffer the streaming responses is the best one. 

One question, do you mean state_stream_response_body() since we are interested in the response?

https://github.com/mitmproxy/mitmproxy/blob/cf362f2e85736d8724523a7925b931d37479f703/mitmproxy/proxy/layers/http/__init__.py#L424-L451

**Reply by mhils** [reply]
This option should work for both requests and responses. So we need to fix up both the request and the response method. :)

**Reply by mkiami** [reply]
I see, right. Thank you! I'm starting to have a look :)

**Comment by Dieken** [comment]
I got a workaround:

```python
from mitmproxy.proxy.utils import ReceiveBuffer

streaming = True

def accumulateData(data, bytes):
    data.content += bytes
    return bytes
 
 class MyAddon:

     def requestheaders(self, flow):
        if streaming:
            #flow.request.data.content = b''
            flow.request.data.content = ReceiveBuffer()     # HACK: more efficient but wrong data type
            flow.request.stream = lambda bytes: accumulateData(flow.request.data, bytes)
 
     def responseheaders(self, flow):
        if streaming:
            #flow.response.data.content = b''
            flow.response.data.content = ReceiveBuffer()    # HACK: more efficient but wrong data type
            flow.response.stream = lambda bytes: accumulateData(flow.response.data, bytes)
 
     def request(self, flow):
        message = flow.request
         if isinstance(message.data.content, ReceiveBuffer):
             message.data.content = bytes(message.data.content)

        # other code here

     def response(self, flow):
        message = flow.response
         if isinstance(message.data.content, ReceiveBuffer):
             message.data.content = bytes(message.data.content)

        # other code here

```
